File: goes/hurst.jl
using Statistics

include("read.jl")

nn = [4, 8, 16, 32, 64, 128]

ti = df[:, :Time]
fl = df[:, :Flux1]

# Estimate the Hurst parameter.
function hurst(ti, fl, dr, d)

	r = zeros(length(dr), 2)
	for (j, m) in enumerate(dr)

		# Generate a matrix of non-overlapping blocks of
		# size m.
		_, flx = make_blocks(ti, fl, m, d)

		# Calculate the sample mean of each block.
		bm = mean(flx, dims=1)[:]

		# Take the sample variance of the block means.
		r[j, :] = [m, var(bm)]
	end

	# Estimate the Hurst exponent from the variances of
	# the sample means.
	rl = log.(r)
	b = cov(rl[:, 1], rl[:, 2]) / var(rl[:, 1])
	a = mean(rl[:, 2]) - b*mean(rl[:, 1])

	return 1 + b/2
end

# As a check, estimate the Hurst parameter for 
# IID normal data (the true value is 1/2). 
fx = randn(length(ti))
h0 = hurst(ti, fx, nn, 0)
println(h0)

# As another check, simulate correlated data
# with short-range dependence (the true value
# is 1/2).
fx = randn(length(ti))
r = 0.5
for i in 2:length(ti)
	fx[i] = r*fx[i-1] + sqrt(1 - r^2)*fx[i]
end
h0 = hurst(ti, fx, nn, 0)
println(h0)

# Estimate the Hurst Parameter for the GOES data.
h0 = hurst(ti, 1e8*fl, nn, 0)
h1 = hurst(ti, 1e8*fl, nn, 1)
h2 = hurst(ti, 1e8*fl, nn, 2)
println([h0, h1, h2])

File: goes/hurst.py
import numpy as np
from read import *

df = get_goes(2017)

nn = [4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048]

# Estimate the Hurst parameter.
def hurst(df, nn, d):

    r = np.zeros((len(nn), 2))
    for j, m in enumerate(nn):

        # Generate a matrix of non-overlapping blocks of
        # size m.
        _, flx = make_blocks(df, m, d)

        # Calculate the sample mean of each block.
        bm = flx.mean(1)

        # Take the sample variance of the block means.
        r[j, :] = [m, bm.var()]

    # Estimate the Hurst exponent from the variances of
    # the sample means.
    rl = np.log(r)
    b = np.cov(rl[:, 0], rl[:, 1])[0, 1] / np.var(rl[:, 0])
    a = rl[:, 1].mean() - b*rl[:, 0].mean()

    return 1 + b/2

# As a check, estimate the Hurst parameter for 
# IID normal data (the true value is 1/2). 
dx = df.iloc[0:100000, :].copy()
dx["Flux1"] = np.random.normal(size=100000)
h0 = hurst(dx, nn, 0)
print("Estimated Hurst parameter for IID standard normal data:")
print(h0)

# As another check, simulate correlated data
# with short-range dependence (the true value
# is 1/2).
fx = np.random.normal(size=dx.shape[0])
r = 0.5
for i in range(1, len(fx)):
    fx[i] = r*fx[i-1] + np.sqrt(1 - r**2)*fx[i]
dx["Flux1"] = fx
h0 = hurst(dx, nn, 0)
print("\nEstimated Hurst parameter for short-range dependent normal data:")
print(h0)

# Estimate the Hurst Parameter for the GOES data.
print("\nEstimated Hurst parameter for blocks of Flux-1 data:")
for dx in np.array_split(df, 20):
    h0 = hurst(dx, nn, 0)
    h1 = hurst(dx, nn, 1)
    print([h0, h1])
